### ISSUE TITLE
Fetch translated exercises

### ISSUE THREAD

**Issue by aronasorman** [learningequality/ka-lite#2729]
Right now we only support officially internationalized KA sites like fr, pt-br and es. We wanna expand that to any user-translated exercises like de. We should also fetch translations for the Perseus exercises.

**Comment by jesumer** [comment]
Hi @aronasorman,

How to replicate this issue? Or how do the fetching translations works?

**Comment by aronasorman** [comment]
@jesumer the `languagepackdownload` management command is the file you're looking for!

**Comment by cpauya** [comment]
Hi, am attaching the email of @jamalex relating to this issue along with the scripts attached for reference.  As @aronasorman said - after due modifications, we may be able to open-source these!

FYI: I can't attach `*.py` or `*.zip` files here so I added it on Google Drive and here's the link to the folder with the scripts: https://drive.google.com/open?id=0Bwf_-YL9LvlVN2lVX1hEeWRRRlE&authuser=0

Start of Email History:

	From: Aron Asor
	Date: Mon, Dec 20, 2014 at 2:10 AM
	This seems pretty tricky. I'll go ahead and read their stuff, and I'll give 
	an assessment in a bit on how hard it is to add this to our pipeline. 
	Best case, we open-source this before they do!


	---------- Forwarded message ----------
	From: James Irwin 
	Date: Mon, Dec 15, 2014 at 7:45 PM
	Subject: Re: Internationalization tools for khan-exercises and Perseus
	To: Jamie Alexandre
	Cc: Ben Eater, Aron Asor, Richard Tibbles

	Hi Jamie,

	Sorry for delayed response.  Unfortunately the scripts to build the
	translated version of the html files are pretty intertwined with the
	rest of our compiler and it is not easy to remove.  I have attached
	exercises/babel.py and kake/translate_exercises.py which uses
	kake/translate_javascript.  We would like to refactor and open source
	this at sometime, but its a substantial amount of work so we can't do
	it soon.  With some reworking you should be able to use the
	translate() function in translate_exercises to create the files for
	the other languages.

	In terms of translating perseus questions you have the right idea.  It
	also has some trickiness though.  I've added our
	assessment_items/models.py file which has a
	traverse_natural_language_parts() method and an
	assessment_items/i18n.py file which has a function
	translate_serialized_assessment_item that will translate the fully
	parsed item.data.  It also translates decimals automatically in case
	translators have not done so.

	Hope this helps and you can get it working for all languages.  Let me
	know if you have any further questions here.

	Best,
	James

	On Tue, Dec 9, 2014 at 11:34 AM, Jamie Alexandre
	<jamie@learningequality.org> wrote:
	> Hey James and Ben!
	>
	> It was great to see you again last week, and to catch up -- the party was a
	> blast! :D
	>
	> I just wanted to follow up about some of the questions we have for
	> internationalization of exercises. For khan-exercises, my understanding
	> (please correct me if wrong) is there are scripts that are run to generate
	> pre-built translated versions. So far, we've just been scraping the
	> generated khan-exercises HTML from the site for languages that are launched,
	> and bundling these for distribution with KA Lite, but we also want to be
	> able to build for languages that aren't yet released on KA. Is it possible
	> for us to access/use the scripts for this build process?
	>
	> For Perseus, my understanding is that each assessment item is loaded into
	> one of the POT files (e.g. learn.math.basic-geo.exercises.pot) and
	> translated on CrowdIn. We should then be able to download and merge these PO
	> files into our master PO/MO file and distribute it with KA Lite, then just
	> pass itemData.question.content and itemData.hints.content through gettext
	> before passing them to Perseus for rendering. Is there anything important we
	> should be aware of in running gettext on Perseus item data?
	>
	> I'm including Aron here, as he's been leading i18n for us, and will be
	> working on integrating exercise translations for our upcoming 0.13.0
	> release, which will bring us back up to date with the latest KA content. :)
	>
	> Thanks!
	> Jamie.

End of Email History

**Comment by aronasorman** [comment]
To expound on this, there needs to be two translation systems, one for khan-exercises and another for Perseus. The difference stems from how the questions are stored.

For khan-exercises, the questions are stored in the static html files. My guess is that the way we wanna translate here is to replace the text statically and write that to the zip, or to a folder and then zipped up.

For Perseus, exercises questions are stored inside `ka-lite/data/khan/assessmentitems.json`. It's a JSON map, with each item's `item_data` containing the data for the question. You'll want to create a function that takes in the `assessmentitems.json` in English and then outputs a new json file that has been translated.


I'm not exactly sure what James' py files contain, but it should provide some insight on how their i18n pipeline works. We might wanna start with tackling the Perseus exercises, since that might be easier, and then open another issue for khan-exercises.

**Comment by jesumer** [comment]
HI @aronasorman,

I have verified the languagepackdownload command and used "de" as a language. I have noticed that the languagepackdowload processes is not working. As we can see, I had already downloaded the de language pack from https://learningequality.org into my local.(see the screenshot)
![screen shot 2015-01-08 at 5 38 12 pm](https://cloud.githubusercontent.com/assets/2810009/5660639/82a27328-975d-11e4-95f9-8b3ed1d18d8a.png)

And then it will supposedly work the processes. In my basis it will supposedly added the "de" folder at ka-lite/kalite/i18n/static/khan-exercises which isn't.(see the screenshot)
![screen shot 2015-01-08 at 5 37 19 pm](https://cloud.githubusercontent.com/assets/2810009/5660695/2d3c9930-975e-11e4-994b-7abe552c6852.png)

**Comment by aronasorman** [comment]
Did you set German as your default language?

**Comment by aronasorman** [comment]
It should've been added into the `ka-lite/locale` directory. Since there are no translated exercises for German, there will be no de folder in khan-exercises.

**Comment by jesumer** [comment]
Hi @aronasorman,

Ah I see. Yeah I set German which is "de".

**Comment by jesumer** [comment]
Findings with Cyril.

  We tackled the contentload command at distributed and we tackled the update_language_packs at the central server. We found out that the json's(exercises.json, topics.json, etc) at kalite/data/khan at distributed was from khan academy api which doesn't do at the central server.
 
  I do the search about how the assessment.json works and end up  [kalite/distributed/static/perseus/get_all_items.py](https://github.com/learningequality/ka-lite/blob/develop/kalite/distributed/static/perseus/get_all_items.py#L33) and playing around the [urls](https://www.khanacademy.org/api/v1/assessment_items/xfc05e430f582a38d?lang=fr) used like changing the language code at the end.

  We propose to create the app which is 
 - collecting the non-translated strings from gettext
 - sync from distributed to central server to collect them as po files which we can use to send into crowdin
 - download the translated strings from crowdin and use it to distributed.

This will be discuss with Cyril for further details about the proposal.

**Comment by cpauya** [comment]
Hi @aronasorman - as per our last talk with @jesumer, here are the things we need to do for this issue:

1. At the central - use `update_language_pack` to download Khan Academy strings from Crowdin to build the "deutsch language pack".
1.  At the distributed - use `languagepackdownload -l de` to download the language pack
1. Check which of the exercise templates / scripts needs to be i18nize so that the translated strings are loaded.

**Comment by jesumer** [comment]
Hi All,

As what my progress yesterday I was able to replicate and determine what are the strings that needs to have internationalized at the perseus exercise as seen in the screenshot.
![screen shot 2015-01-22 at 9 22 09 am](https://cloud.githubusercontent.com/assets/2810009/5848890/b3e4beee-a218-11e4-90e8-0459188d7732.png)

And here is where I planned to get the strings from Asssesment Item:
![screen shot 2015-01-22 at 9 19 38 am](https://cloud.githubusercontent.com/assets/2810009/5848911/fac3cc38-a218-11e4-9ea4-0daa2e049f44.png)

Now the problem is the assessment item has special characters on it which I think it was important. Here is the sample item json data:
![screen shot 2015-01-22 at 9 18 33 am](https://cloud.githubusercontent.com/assets/2810009/5848933/496f726a-a219-11e4-8c2e-071c385ce1f1.png)
 
Cyril suggest to put "gettext()" into the template in which I figure out where it was happen or where should I find that template.

**Comment by cpauya** [comment]
@jesumer Let's take a look at the scripts that @jamalex forwarded as @aronasorman suggested.

**Comment by jesumer** [comment]
Hi All,

I was able to use the script(i18n.py) script from James at Khan Academy.
I noticed of using this import modules:
 1. import api.jsonify
    - Which used on jsonify.as_serializable(assessment_item) at def is_fully_translated() function.
 2. from intl import i18n
    - Which uses i18n._(text) at def _maybe_translate() function, i18n.request_language_decimal_format(), i18n.format_decimal() at def _translate_number() function
 3. import intl.regexps
    - Which uses regexps.NO_NEED_TO_TRANSLATE.match(text)
 4. import intl.request
    - Which uses request.locale_for_mo(), request.jipt_locale_for_mo().

Some of the function are important like i18n._(text), i18n.request_language_decimal_format(), i18n.format_decimal(). I need to know where I can get this. Thank you.

**Comment by jesumer** [comment]
Updates: 

I pulled the develop branch update and successfully run the perseus exercises. I tried wrapped it like _(ASSESSMENT_ITEMS) but that doesn't work and get errors because of some special characters inside like '\r\n', \n', '\r', and '\n' so Now, I started coding the assessment json to have it translatable or wrapping the strings into _() function.

Jesumer

**Comment by jesumer** [comment]
Updates and findings:

I have tried to find out the sample German strings that has been approved from crowdin.
![screen shot 2015-02-02 at 4 10 22 pm](https://cloud.githubusercontent.com/assets/2810009/5997516/9ee6e5b6-aaf7-11e4-87b5-0d8f8ca7b1ed.png)
and this is for french language
![screen shot 2015-02-02 at 4 42 33 pm](https://cloud.githubusercontent.com/assets/2810009/5997733/5e0ea01c-aafa-11e4-808d-3fd947bd9856.png)


I set the german language and this is the output into our site:
![screen shot 2015-02-02 at 4 11 58 pm](https://cloud.githubusercontent.com/assets/2810009/5997570/6fb23be6-aaf8-11e4-86d4-1f2b6eb1816c.png)
And for the french:
![screen shot 2015-02-02 at 4 42 15 pm](https://cloud.githubusercontent.com/assets/2810009/5997735/6f155766-aafa-11e4-8c91-6a9cfee634da.png)


I also tried to find out in the python shell. Here is the german language sample:
![screen shot 2015-02-02 at 4 56 10 pm](https://cloud.githubusercontent.com/assets/2810009/5997900/233d7bf0-aafc-11e4-9f9b-41f7ac363423.png)

And the french:
![screen shot 2015-02-02 at 4 58 47 pm](https://cloud.githubusercontent.com/assets/2810009/5997935/7cf625f2-aafc-11e4-95b7-828555fe5220.png)

 Then after the "languagepackdownload" in distributed, I check out and find the strings(e.g "Create a picture graph to show how many teeth each student has lost.") in the mo file located at ka-lite/locale/de/LC_MESSAGES/ and also at ka-lite/locale/fr/LC_MESSAGES/ but unfortunately it isn't there and the string doesn't exist. Is there anything that I missed?. Base of my understanding, the mo files from central server has all the translated strings accordingly to our language we set. I also have the python scripts to enclose the non-translated string from our assessment items into gettext or _().


Best,
Jesumer

**Comment by cpauya** [comment]
Hi @jesumer - you must copy the whole untranslated string from Crowdin (not portion only) and use that on your Python shell.  Example based on your link above:

https://crowdin.com/translate/khanacademy/27617/enus-de

```
**Create a picture graph to show how many seeds Johnny Appleseed planted in each location.**\n\nLocation | Apple seeds \n- | :-: | -\nCharleston | $2$ \nLouisville | $6$ \nRichmond | $5$  \nSpringfield | $6$  \n\n![](https://ka-perseus-images.s3.amazonaws.com/a5b68232872a3e078a942f9c298e815b2a92f4e9.png)\n\n[[☃ plotter 1]]\n
```

**Comment by cpauya** [comment]
@jesumer I have just filed a related issue at https://github.com/fle-internal/ka-lite-central/issues/225 on central repo.

Please verify if that indeed affects/fixes this issue.

**Comment by jesumer** [comment]
Hi,

We have already downloaded the translated po files at the central server
we can replicate this in our distributed  console and have it translated.
Here is the screenshot
![screen shot 2015-02-04 at 9 00 33 pm](https://cloud.githubusercontent.com/assets/2810009/6040768/9fb1134e-acb0-11e4-8995-62b84751462a.png)

This is the po files from crowdin and search the translated string("Telling time without labels")"
![screen shot 2015-02-04 at 9 01 50 pm](https://cloud.githubusercontent.com/assets/2810009/6040787/c366e804-acb0-11e4-8a0c-ad49b645582e.png)

Now the problem is we can't find the string("Telling time without labels") into the distributed browser. I think the assessment items json is the problem.

Here is the topic tree 
![screen shot 2015-02-04 at 9 06 51 pm](https://cloud.githubusercontent.com/assets/2810009/6040866/8f4c7e66-acb1-11e4-82f7-49642bd78830.png)

And here is the result at the browser
![screen shot 2015-02-04 at 9 06 59 pm](https://cloud.githubusercontent.com/assets/2810009/6040868/91427c20-acb1-11e4-84dc-53f5e426c006.png)


Jesumer

**Comment by jesumer** [comment]
Hi,

I'm working with the multiple hints.

Jesumer

**Comment by jesumer** [comment]
Hi,

I'm done with the multiple hints.

Jesumer

**Comment by jesumer** [comment]
Hi all,

I have  already refactored my script at get_assessment_item_data. We can now use it.
Thanks

Jesumer

**Comment by jesumer** [comment]
Hi,

After running the perseus exercise that is default to de language code, here is my example working scripts at the browser.
![screen shot 2015-02-05 at 2 48 20 pm](https://cloud.githubusercontent.com/assets/2810009/6119857/7486271a-b109-11e4-82f9-a01287e65d20.png)

Notice the Answer pane at the right. It has a translated string to German language.

Jesumer

**Comment by alainschaefer** [comment]
DE is still working on the review and approval for geo, I recommend you use
early-math exercises for testing as there you'll have already 95% of the
strings translated and approved.

**Comment by arceduardvincent** [comment]
Hi @jesumer I will review this issue you made and the `PR` you made so this issue will be close.

**Comment by aronasorman** [comment]
HOLD UP! No one close this issue yet.

**Comment by MCGallaspy** [comment]
@aronasorman what remains before closing this issue?

**Comment by aronasorman** [comment]
I just merged it to make it easier to test the dummy language packs. Once the tests are fixed for that PR we can merge that, and everyone can test i18n.

**Comment by aronasorman** [comment]
With both dummy language packs and perseus exercises merged in, we can now properly test this issue.

I created a dummy language pack by running:
`bin/kalite manage create_dummy_language_pack`

I then switched to Esperanto (the name for the dummy language), and got this:
![screen shot 2015-03-18 at 6 04 20 pm](https://cloud.githubusercontent.com/assets/191955/6722576/79802944-cd99-11e4-9de6-16ed710eea2d.png)

So the interface is partially translated. However, when I open the exercises, I don't see any translated strings:
![screen shot 2015-03-18 at 6 03 52 pm](https://cloud.githubusercontent.com/assets/191955/6722585/8d72a972-cd99-11e4-9d10-06630a0b3a92.png)

**Comment by aronasorman** [comment]
It seems like the strings aren't in the fetched po file in the first place.

**Comment by aronasorman** [comment]
Ah, seems like there's no `en` language in Khan Academy, so it might not be a good base language pack after all.

**Comment by aronasorman** [comment]
String is in `django.po`, but it's not getting substituted in the exercise:
![screen shot 2015-03-18 at 6 25 05 pm](https://cloud.githubusercontent.com/assets/191955/6722753/25f00f76-cd9c-11e4-8b74-78c4fd8d3c50.png)

**Comment by aronasorman** [comment]
Works in the terminal though:
![screen shot 2015-03-18 at 6 26 41 pm](https://cloud.githubusercontent.com/assets/191955/6722767/6123422a-cd9c-11e4-922c-8151a17584bc.png)

**Comment by aronasorman** [comment]
Turns out I just needed to refetch the assessment items. I'm now getting this:
![screen shot 2015-03-19 at 9 18 17 am](https://cloud.githubusercontent.com/assets/191955/6735039/eb39b90e-ce18-11e4-93f1-6f698b0e26bd.png)

Notice the accented instructions on the right side, but unaccented strings on the question area.

**Comment by aronasorman** [comment]
Hints are translated:
![screen shot 2015-03-19 at 9 19 39 am](https://cloud.githubusercontent.com/assets/191955/6735077/1c47b672-ce19-11e4-81f8-4ed4c5ec8b7e.png)

**Comment by aronasorman** [comment]
Ok, I think I found the issue. Django's `gettext` can't seem to find the translated string, even though they're in the po file and they look exactly the same.

See this entry in the po file, with accents and all:
![screen shot 2015-03-19 at 9 34 31 am](https://cloud.githubusercontent.com/assets/191955/6735426/4c939db2-ce1b-11e4-8e24-291f3635effd.png)

But when I go to pdb, `ugettext` can't find it:
![screen shot 2015-03-19 at 9 34 17 am](https://cloud.githubusercontent.com/assets/191955/6735437/5f89883c-ce1b-11e4-9a4b-899f08e69ac0.png)

**Comment by jamalex** [comment]
Looks like it could be an issue of how the pieces are being chunked up? Are we using that code that KA sent us?

**Comment by aronasorman** [comment]
@jamalex Yeah, all the exercise-related strings are from Khan Academy. They might be using a different system for localization?

**Comment by jamalex** [comment]
Right, the strings are from them. But they shared code with us that parsed `item_data` into the translatable chunks or something, right? The po files don't just contain the entire contents of each `item_data` field, do they?

**Comment by jamalex** [comment]
Looks like we just do: `answerarea_content = _(answerarea_content)`

Is that what KA does?

**Comment by rtibbles** [comment]
How would that work when we have changed the URLs, btw?

**Comment by jamalex** [comment]
> How would that work when we have changed the URLs, btw?

Magic! https://github.com/learningequality/ka-lite/pull/3342/files#diff-7f19457f4d6f7fd73961c0d5d92a1dd5R132

**Comment by aronasorman** [comment]
I think the issue here is how python's `gettext` finds strings, as polib can find the exact same strings (and thus fetch the translations) while `_` can't:

![screen shot 2015-03-19 at 10 18 42 am](https://cloud.githubusercontent.com/assets/191955/6736331/5b1de74c-ce21-11e4-82e8-ed42a4b0b5ad.png)

**Comment by aronasorman** [comment]
@jamalex we didn't get too much use out of KA's files, as we needed (i think) 5 more modules. I believe it's too tied to their code.

**Comment by aronasorman** [comment]
Took a bit of digging, but I finally found a dict that maps a string to its translated counterpart, as read by python's `gettext`:
```python
from django.utils import translation
translation.activate('eo') # "eo" is the code for the DEBUG language.
catalog = translation.trans_real.catalog()._catalog
```
`catalog` is what you're looking for.

**Comment by aronasorman** [comment]
So, it turns out to be an encoding issue.

Django reads in PO files as `ascii`:
```
'**Which choices represent  the number $550$ ?**\n\n[[\xe2\x98\x83 radio 1]]'
```

However, we read in assessment items as unicode:
```
u'**Which choices represent  the number $550$ ?**\n\n[[\u2603 radio 1]]'
```

The solution is to encode the untranslated strings into ascii before passing them to the `gettext` function:
```python
        question_content = _(item_data['question']['content'].encode('utf-8'))
```

**Comment by aronasorman** [comment]
And thus we get:
![screen shot 2015-03-19 at 11 26 54 am](https://cloud.githubusercontent.com/assets/191955/6737762/ed4d644a-ce2a-11e4-83c7-26aeca8d5677.png)

**Comment by rtibbles** [comment]
Wooooot!

**Comment by aronasorman** [comment]
So I got this specific exercise translated:

![screen shot 2015-03-19 at 12 38 30 pm](https://cloud.githubusercontent.com/assets/191955/6739280/eb90f4d2-ce34-11e4-944e-58c142a3db94.png)


However, it looks like I'm gonna have to go through each type of exercise, find their structure, and translate them.

**Comment by aronasorman** [comment]
Just created a function that will automatically translate all types of questions \o/.

While waiting on #3342, I'll work on getting the other text translated. These are most likely related to backbone.js + `djangojs` issues.

**Comment by aronasorman** [comment]
Note: looks like a a majority of the health and medicine topics don't even have po entries in them:
![screen shot 2015-03-19 at 4 26 45 pm](https://cloud.githubusercontent.com/assets/191955/6743267/e028717c-ce54-11e4-8c49-5bf0936782a5.png)

**Comment by jamalex** [comment]
> Note: looks like a a majority of the health and medicine topics don't even have po entries in them:

Well that's silly! We should ask KA
